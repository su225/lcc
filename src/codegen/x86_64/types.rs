use std::fmt::{Display, Formatter};
use crate::codegen::x86_64::register::Register;
use crate::tacky::TackySymbol;

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct StackOffset(pub(crate) isize);

#[derive(Debug, Clone, PartialEq)]
pub enum AsmOperand {
    Imm32(i32),
    Reg(Register),
    Pseudo(TackySymbol),
    Stack { offset: StackOffset },
}

impl Display for AsmOperand {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", match &self {
            AsmOperand::Imm32(n) => format!("${}", n),
            AsmOperand::Reg(r) => r.to_string(),
            AsmOperand::Pseudo(p) => format!("<<{}>>", p),
            AsmOperand::Stack { offset } => format!("{}(%rbp)", offset.0),
        })
    }
}

#[derive(Debug, PartialEq)]
pub enum AsmInstruction {
    AllocateStack(usize),
    Mov8 { src: AsmOperand, dst: AsmOperand },
    Mov32 { src: AsmOperand, dst: AsmOperand },
    Neg32 { op: AsmOperand },
    Not32 { op: AsmOperand },
    Add32 { src: AsmOperand, dst: AsmOperand },
    Sub32 { src: AsmOperand, dst: AsmOperand },
    IMul32 { src: AsmOperand, dst: AsmOperand },
    IDiv32 { divisor: AsmOperand },
    And32 { src: AsmOperand, dst: AsmOperand },
    Or32 { src: AsmOperand, dst: AsmOperand },
    Xor32 { src: AsmOperand, dst: AsmOperand },
    Sal32 { src: AsmOperand, dst: AsmOperand },
    Sar32 { src: AsmOperand, dst: AsmOperand },
    Shl32 { src: AsmOperand, dst: AsmOperand },
    Shr32 { src: AsmOperand, dst: AsmOperand },
    SignExtendTo64, // Sign extend %eax to 64-bit into %edx
    Ret,
}

#[derive(Debug, PartialEq)]
pub struct AsmFunction {
    pub name: TackySymbol,
    pub instructions: Vec<AsmInstruction>,
}

#[derive(Debug, PartialEq)]
pub struct AsmProgram {
    pub functions: Vec<AsmFunction>,
}