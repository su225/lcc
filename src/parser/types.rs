use derive_more::Add;
use serde::Serialize;
use crate::common::{Location, Radix};

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct Symbol<'a> {
    pub name: &'a str,
    pub(crate) location: Location,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum UnaryOperator {
    Complement,
    Negate,
    Not,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum BinaryOperatorAssociativity {
    Left, Right
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum BinaryOperator {
    Add,
    Subtract,
    Multiply,
    Divide,
    Modulo,

    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
    LeftShift,
    RightShift,

    And,
    Or,
    Equal,
    NotEqual,
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
}

#[derive(Debug, PartialEq, Ord, PartialOrd, Eq, Add, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct BinaryOperatorPrecedence(pub(crate) u16);

impl BinaryOperator {
    #[inline]
    pub(crate) fn associativity(&self) -> BinaryOperatorAssociativity {
        match self {
            BinaryOperator::Add => BinaryOperatorAssociativity::Left,
            BinaryOperator::Subtract => BinaryOperatorAssociativity::Left,
            BinaryOperator::Multiply => BinaryOperatorAssociativity::Left,
            BinaryOperator::Divide => BinaryOperatorAssociativity::Left,
            BinaryOperator::Modulo => BinaryOperatorAssociativity::Left,
            BinaryOperator::BitwiseAnd => BinaryOperatorAssociativity::Left,
            BinaryOperator::BitwiseOr => BinaryOperatorAssociativity::Left,
            BinaryOperator::BitwiseXor => BinaryOperatorAssociativity::Left,
            BinaryOperator::LeftShift => BinaryOperatorAssociativity::Left,
            BinaryOperator::RightShift => BinaryOperatorAssociativity::Left,
            BinaryOperator::And => BinaryOperatorAssociativity::Left,
            BinaryOperator::Or => BinaryOperatorAssociativity::Left,
            BinaryOperator::Equal => BinaryOperatorAssociativity::Left,
            BinaryOperator::NotEqual => BinaryOperatorAssociativity::Left,
            BinaryOperator::LessThan => BinaryOperatorAssociativity::Left,
            BinaryOperator::LessThanOrEqual => BinaryOperatorAssociativity::Left,
            BinaryOperator::GreaterThan => BinaryOperatorAssociativity::Left,
            BinaryOperator::GreaterThanOrEqual => BinaryOperatorAssociativity::Left,
        }
    }

    #[inline]
    pub(crate) fn precedence(&self) -> BinaryOperatorPrecedence {
        match self {
            BinaryOperator::Multiply
            | BinaryOperator::Divide
            | BinaryOperator::Modulo => BinaryOperatorPrecedence(50),

            BinaryOperator::Add | BinaryOperator::Subtract => BinaryOperatorPrecedence(45),
            BinaryOperator::LeftShift | BinaryOperator::RightShift => BinaryOperatorPrecedence(42),

            BinaryOperator::LessThan
            | BinaryOperator::LessThanOrEqual
            | BinaryOperator::GreaterThan
            | BinaryOperator::GreaterThanOrEqual => BinaryOperatorPrecedence(40),

            BinaryOperator::Equal | BinaryOperator::NotEqual => BinaryOperatorPrecedence(38),

            BinaryOperator::BitwiseAnd => BinaryOperatorPrecedence(36),
            BinaryOperator::BitwiseXor => BinaryOperatorPrecedence(34),
            BinaryOperator::BitwiseOr => BinaryOperatorPrecedence(32),

            BinaryOperator::And => BinaryOperatorPrecedence(30),
            BinaryOperator::Or => BinaryOperatorPrecedence(28),
        }
    }
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ExpressionKind<'a> {
    IntConstant(&'a str, Radix),
    Unary(UnaryOperator, Box<Expression<'a>>),
    Binary(BinaryOperator, Box<Expression<'a>>, Box<Expression<'a>>),
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct Expression<'a> {
    pub(crate) location: Location,
    pub kind: ExpressionKind<'a>,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum PrimitiveKind {
    Integer,
    UnsignedInteger,
    LongInteger,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum TypeExpressionKind {
    Primitive(PrimitiveKind),
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct TypeExpression {
    pub location: Location,
    pub kind: TypeExpressionKind,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum StatementKind<'a> {
    Return(Expression<'a>),
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct Statement<'a> {
    pub location: Location,
    pub kind: StatementKind<'a>,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct FunctionDefinition<'a> {
    pub location: Location,
    pub name: Symbol<'a>,
    pub body: Vec<Statement<'a>>,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct ProgramDefinition<'a> {
    pub functions: Vec<FunctionDefinition<'a>>,
}