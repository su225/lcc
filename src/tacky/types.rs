use derive_more::Display;
use crate::parser::types::{BinaryOperator, UnaryOperator};

pub(crate) const COMPILER_GEN_PREFIX: &'static str = "<t>";

#[derive(Debug, Eq, PartialEq, Clone, Hash, Display)]
pub struct TackySymbol(pub String);

impl From<&str> for TackySymbol {
    fn from(value: &str) -> Self {
        TackySymbol(value.to_string())
    }
}

impl TackySymbol {
    pub fn is_generated(&self) -> bool {
        self.0.starts_with(COMPILER_GEN_PREFIX)
    }
}

#[derive(Debug, PartialEq)]
pub struct TackyProgram {
    pub functions: Vec<TackyFunction>,
}

#[derive(Debug, PartialEq)]
pub struct TackyFunction {
    pub identifier: TackySymbol,
    pub body: Vec<TackyInstruction>,
}

#[derive(Debug, PartialEq)]
pub enum TackyUnaryOperator {
    Complement,
    Negate,
    Not,
}

impl From<&UnaryOperator> for TackyUnaryOperator {
    fn from(value: &UnaryOperator) -> Self {
        match value {
            UnaryOperator::Complement => TackyUnaryOperator::Complement,
            UnaryOperator::Negate => TackyUnaryOperator::Negate,
            UnaryOperator::Not => todo!(),
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum TackyBinaryOperator {
    Add,
    Subtract,
    Multiply,
    Divide,
    Modulo,
    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
    LeftShift,
    RightShift,
    Equal,
    NotEqual,
    LessThan,
    LessOrEqual,
    GreaterThan,
    GreaterOrEqual,
}

impl From<&BinaryOperator> for TackyBinaryOperator {
    fn from(value: &BinaryOperator) -> Self {
        match value {
            BinaryOperator::Add => TackyBinaryOperator::Add,
            BinaryOperator::Subtract => TackyBinaryOperator::Subtract,
            BinaryOperator::Multiply => TackyBinaryOperator::Multiply,
            BinaryOperator::Divide => TackyBinaryOperator::Divide,
            BinaryOperator::Modulo => TackyBinaryOperator::Modulo,

            BinaryOperator::BitwiseAnd => TackyBinaryOperator::BitwiseAnd,
            BinaryOperator::BitwiseOr => TackyBinaryOperator::BitwiseOr,
            BinaryOperator::BitwiseXor => TackyBinaryOperator::BitwiseXor,
            BinaryOperator::LeftShift => TackyBinaryOperator::LeftShift,
            BinaryOperator::RightShift => TackyBinaryOperator::RightShift,

            _ => todo!(),
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum TackyValue {
    Constant32(i32),
    Variable(TackySymbol),
}

#[derive(Debug, PartialEq)]
pub enum TackyInstruction {
    Unary {
        operator: TackyUnaryOperator,
        src: TackyValue,
        dst: TackyValue,
    },
    Binary {
        operator: TackyBinaryOperator,
        src1: TackyValue,
        src2: TackyValue,
        dst: TackyValue,
    },
    Return(TackyValue),
    Copy { src: TackyValue, dst: TackyValue },
    Jump { target: TackySymbol },
    JumpIfZero { condition: TackyValue, target: TackySymbol },
    JumpIfNotZero { condition: TackyValue, target: TackySymbol },
    Label(TackySymbol),
}