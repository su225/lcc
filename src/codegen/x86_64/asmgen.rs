use std::io;
use std::io::Write;
use crate::codegen::x86_64::types::*;
use crate::codegen::x86_64::types::AsmInstruction::*;

pub fn emit<W: Write>(asm_code: AsmProgram, mut w: W) -> io::Result<()> {
    for f in asm_code.functions.into_iter() {
        emit_function(f, &mut w)?;
    }
    Ok(())
}

fn emit_function<W: Write>(f: AsmFunction, w: &mut W) -> io::Result<()> {
    w.write_fmt(format_args!("    .globl _{function_name}\n", function_name = f.name))?;
    w.write_fmt(format_args!("_{function_name}:\n", function_name = f.name))?;
    emit_function_prologue(w)?;
    for instr in f.instructions {
        emit_instruction(&instr, w)?;
    }
    Ok(())
}

macro_rules! emit_instruction {
    ($w:expr, $($arg:tt)*) => {
        writeln!($w, "    {}", format!($($arg)*))
    };
}

fn emit_instruction<W: Write>(instr: &AsmInstruction, w: &mut W) -> io::Result<()> {
    match instr {
        Mov8 { src, dst } => emit_instruction!(w, "movb {src}, {dst}")?,
        Mov32 { src, dst } => emit_instruction!(w, "movl {src}, {dst}")?,
        Ret => {
            emit_function_epilogue(w)?;
            emit_instruction!(w, "ret")?
        },
        Not32 { op } => emit_instruction!(w, "notl {op}")?,
        Neg32 { op } => emit_instruction!(w, "negl {op}")?,
        AllocateStack(stack_size) => emit_instruction!(w, "subq ${stack_size}, %rsp")?,
        Add32 { src, dst } => emit_instruction!(w, "addl {src}, {dst}")?,
        Sub32 { src, dst } => emit_instruction!(w, "subl {src}, {dst}")?,
        IMul32 { src, dst } => emit_instruction!(w, "imull {src}, {dst}")?,
        IDiv32 { divisor } => emit_instruction!(w, "idivl {divisor}")?,
        SignExtendTo64 => emit_instruction!(w, "cdq")?,

        And32 { src, dst } => emit_instruction!(w, "andl {src}, {dst}")?,
        Or32 { src, dst } => emit_instruction!(w, "orl {src}, {dst}")?,
        Xor32 { src, dst } => emit_instruction!(w, "xorl {src}, {dst}")?,
        Shl32 { src, dst } => emit_instruction!(w, "shll {src}, {dst}")?,
        Shr32 { src, dst } => emit_instruction!(w, "shrl {src}, {dst}")?,
        Sal32 { src, dst } => emit_instruction!(w, "sall {src}, {dst}")?,
        Sar32 { src, dst } => emit_instruction!(w, "sarl {src}, {dst}")?,
    };
    Ok(())
}

fn emit_function_prologue<W: Write>(w: &mut W) -> io::Result<()> {
    emit_instruction!(w, "pushq %rbp")?;
    emit_instruction!(w, "movq  %rsp, %rbp")?;
    Ok(())
}

fn emit_function_epilogue<W: Write>(w: &mut W) -> io::Result<()> {
    emit_instruction!(w, "movq %rbp, %rsp")?;
    emit_instruction!(w, "popq %rbp")?;
    Ok(())
}