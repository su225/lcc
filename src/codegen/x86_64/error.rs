use std::num::ParseIntError;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum CodegenError {
    #[error(transparent)]
    IntImmediateParseError(#[from] ParseIntError),
}