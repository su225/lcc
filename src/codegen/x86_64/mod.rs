pub mod codegen;
pub mod asmgen;
pub mod register;

pub use codegen::*;
pub use asmgen::*;