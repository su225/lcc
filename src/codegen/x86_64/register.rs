use std::fmt::{Display, Formatter};
use crate::codegen::x86_64::register::Register::*;

#[derive(Debug, Clone, PartialEq)]
pub enum Register {
    EAX, EDX, R10D, R11D,

    RSP, RBP,

    R10B, AL, CL,
}

impl Display for Register {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "%{}", match &self {
            EAX => "eax",
            EDX => "edx",
            R10D => "r10d",
            R11D => "r11d",

            RSP => "rsp",
            RBP => "rbp",

            R10B => "r10b",
            CL => "cl",
            AL => "al",
        })
    }
}