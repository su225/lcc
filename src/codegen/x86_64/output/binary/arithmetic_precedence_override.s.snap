---
source: src/codegen/x86_64/asmgen.rs
description: binary operators
expression: asm_output
info: binary/arithmetic_precedence_override.c
---
    .globl _main
_main:
    pushq %rbp
    movq  %rsp, %rbp
    subq $16, %rsp
    movl $1, -8(%rbp)
    addl $3, -8(%rbp)
    movl $2, -16(%rbp)
    movl -8(%rbp), %r10d
    movl -16(%rbp), %r11d
    imull %r10d, %r11d
    movl %r11d, -16(%rbp)
    movl -16(%rbp), %eax
    movq %rbp, %rsp
    popq %rbp
    ret
